### PR TITLE
PP-4993 Introduce StripeTransferRequest objects

### PR DESCRIPTION
**Pull request by mrlumbu**
Introduces StripeTransfer{In,Out}Request.
In means a transfer of money from a Stripe Connect account to the Pay platform account.
Out means a transfer of money from the Pay platform account to a Stripe Connect account.

This PR  does not actually use these objects.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                request.getChargeExternalId(),
+                stripeGatewayConfig
+        );
+    }
+
+    @Override
+    public GatewayOrder getGatewayOrder() {
+        List<BasicNameValuePair> params = getCommonPayloadParameters();
+        params.add(new BasicNameValuePair("destination", stripeGatewayConfig.getPlatformAccountId()));
+        params.add(new BasicNameValuePair("transfer_group", transferGroup));
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequest.java:51]: Is our plan to create a charge with `transfer_group` and use it for all subsequent transactions ?

Stripe docs says

> TRANSFERS WITHOUT CHARGES
To credit a connected account without an associated charge, simply make a transfer without a transfer group.

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import uk.gov.pay.connector.gatewayaccount.model.GatewayAccountEntity;
+
+import java.util.List;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static javax.ws.rs.core.MediaType.APPLICATION_FORM_URLENCODED_TYPE;
+
+public class StripeTransferOutRequest extends StripeTransferRequest {
+
+    public StripeTransferOutRequest(String amount,
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequest.java:18]: Can this be `private` ?

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferRequest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    public abstract GatewayOrder getGatewayOrder();
+
+    @Override
+    public URI getUrl() {
+        return URI.create(stripeGatewayConfig.getUrl() + "/v1/transfers");
+    }
+
+    protected List<BasicNameValuePair> getCommonPayloadParameters() {
+        List<BasicNameValuePair> params = new ArrayList<>();
+        params.add(new BasicNameValuePair("amount", amount));
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferRequest.java:37]: Don't we need `transfer_group` here ?

File: src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequestTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        assertThat(stripeTransferInRequest.getUrl(), is(URI.create(stripeBaseUrl + "/v1/transfers")));
+    }
+
+    @Test
+    public void shouldCreateCorrectPayload() {
+        String payload = stripeTransferInRequest.getGatewayOrder().getPayload();
+        
+        assertThat(payload, containsString("destination=" + stripePlatformAccountId));
+        assertThat(payload, containsString("amount=" + refundAmount));
+        assertThat(payload, containsString("transfer_group=" + chargeExternalId));
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequestTest.java:78]: Can we check `currency` too ? Fine to not verify `expand[]` properties if there is no real use case, but would be good to have for test completeness

File: src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequestTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    public void shouldCreateCorrectUrl() {
+        assertThat(stripeTransferOutRequest.getUrl(), is(URI.create(stripeBaseUrl + "/v1/transfers")));
+    }
+
+    @Test
+    public void shouldCreateCorrectPayload() {
+        String payload = stripeTransferOutRequest.getGatewayOrder().getPayload();
+
+        assertThat(payload, containsString("destination=" + stripeConnectAccountId));
+        assertThat(payload, containsString("amount=" + netTransferAmount));
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequestTest.java:70]: Same as above for `currency`

**Review by kbottla — COMMENTED** [review]
Looks good to me. Added few minor comments about usage of `transfer_group` and additional checks in tests.

On a completely separate note, I have noticed that we throw runtime exception when `stripe_account_id` is not available within `gateway_credentials` https://github.com/alphagov/pay-connector/blob/619318b45967c0fbbf37d70a5dce6a00ac8d3058/src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeRequest.java#L26

Would probably be useful to log message (but not in this PR) including `gateway_account_id` so debugging can be easier in such cases.

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequest.java (hunk 1 above)
Comment (mrlumbu) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequest.java:51]: We set the transfer group of this transfer in request to make sure it is linked to the refund. See below comment for why we don't set this on transfer out request

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferRequest.java (hunk 3 above)
Comment (mrlumbu) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferRequest.java:37]: No. On a transfer out request we set `source_transaction` which automatically sets the transfer group to be the same as the original charge. On a transfer in request this is not possible ( for weird stripe reasons) so we have to explicitly set the `transfer_group`.

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequest.java (hunk 2 above)
Comment (mrlumbu) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequest.java:18]: 👍

File: src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequestTest.java (hunk 4 above)
Comment (mrlumbu) [reply src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequestTest.java:78]: 👍

File: src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequestTest.java (hunk 5 above)
Comment (mrlumbu) [reply src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequestTest.java:70]: Member
👍

**Comment by mrlumbu** [comment]
@kbottla thanks! Have addressed your comments

**Review by kbottla — APPROVED** [review]

**Pull request merged**